src/LinuxTracing/LeafFunctionCallManager.h

```diff
@@ -35,7 +36,156 @@ class LeafFunctionCallManager {
   // one in libunwindstack, to match the format of perf_event_open.
   virtual orbit_grpc_protos::Callstack::CallstackType PatchCallerOfLeafFunction(
       const CallchainSamplePerfEventData* event_data, LibunwindstackMaps* current_maps,
-      LibunwindstackUnwinder* unwinder);
+      LibunwindstackUnwinder* unwinder) {
+    return PatchCallerOfLeafFunctionImpl(event_data, current_maps, unwinder);
+  }
+
+  virtual orbit_grpc_protos::Callstack::CallstackType PatchCallerOfLeafFunction(
+      const SchedWakeupWithCallchainPerfEventData* event_data, LibunwindstackMaps* current_maps,
+      LibunwindstackUnwinder* unwinder) {
+    return PatchCallerOfLeafFunctionImpl(event_data, current_maps, unwinder);
+  }
+
+  virtual orbit_grpc_protos::Callstack::CallstackType PatchCallerOfLeafFunction(
+      const SchedSwitchWithCallchainPerfEventData* event_data, LibunwindstackMaps* current_maps,
+      LibunwindstackUnwinder* unwinder) {
+    return PatchCallerOfLeafFunctionImpl(event_data, current_maps, unwinder);
+  }
+
+ private:
+  // Let's unwind one frame using libunwindstack. With that unwinding step, the registers will get
+  // updated and we can detect if $rbp was modified.
+  // (1) If $rbp did not change: We are in a leaf function, which has not modified $rbp. The
+  // leaf's
+  //     caller is missing in the callchain and needs to be patched in. The updated $rip (pc) from
+  //     the unwinding step contains the leaf's caller.
+  // (2) If $rbp was modified, this can either be:
+  //     (a) We are in a non-leaf function and the callchain is already correct.
+  //     (b) We are in a leaf function that modified $rbp. The complete callchain is broken and
+  //     should
+  //         be reported as unwinding error.
+  // As libunwindstack does not report us the canonical frame address (CFA) from an unwinding
+  // step, we cannot differentiate between (2a) and (2b) reliably. However, we do perform the
+  // following validity checks (for the reasoning remember that the stack grows downwards): (I) If
+  // the CFA is computed using $rbp + 16, we know the $rbp was correct, i.e. case (2a) (II)  If
+  // $rbp is below $rsp, $rbp is not a frame pointer, i.e. case (2b) (III) If $rbp moves up the
+  // stack after unwinding, the sampled $rbp is not a frame pointer (2b)
+  //
+  // Note that we cannot simply set libunwindstack to unwind always two frames and compare the
+  // outer frame with the respective one in the callchain carried by the perf_event_open event, as
+  // in case of uprobes overriding the return addresses, both addresses would be identical even if
+  // the actual addresses (after uprobe patching) are not. More (internal) documentation on this
+  // in: go/stadia-orbit-leaf-frame-pointer
+  template <typename CallchainPerfEventDataT>
+  orbit_grpc_protos::Callstack::CallstackType PatchCallerOfLeafFunctionImpl(
```

Comment:
Completely optional, but since we talked about this: as we know the three possible types (`CallchainSamplePerfEventData`, `SchedWakeupWithCallchainPerfEventData`, `SchedSwitchWithCallchainPerfEventData`), if you want to keep the definition of this in the .cpp, you only need to add the three template instantiations in the .cpp, e.g.:
```
template orbit_grpc_protos::Callstack::CallstackType PatchCallerOfLeafFunctionImpl<CallchainSamplePerfEventData>(...);
```

Reply:
Done.